[Staging Mode] Plumb original "viable" op content to resubmit (having the handles to real objects to bind) via wrapped localOpMetadata

markfields opened the pull request:
## Description

Shove the original "viable" op contents onto localOpMetadata so it's available on resubmit, so when exiting staging mode the handle binding works.

### Original approach that did not work:

By holding a queue of pending ops in SharedObjectCore, we can use the original op contents with viable handles when finally resubmitting upon exiting Staging Mode.

This also gives us an easy way to express dirty/saved state on a per-channel basis.

markfields committed:
Some notes

markfields committed:
Keep pending ops in SharedObjectCore

- Enables a isDirty prop (could add events later)
- Allows us to use the viable content, with viable handles (where attachGraph actually does something)

anthony-murphy commented on packages/dds/shared-object-base/src/sharedObject.ts:
```diff
@@ -463,9 +463,20 @@ export abstract class SharedObjectCore<
 
+	private readonly pendingLocalMessages: { content: unknown; localOpMetadata: unknown }[] = [];
+
+	public get isDirty(): boolean {
```
an idea i had was to push message serialization into the ChannelDeltaConnection, and handle serialization, and real message tracking there, as this is also available to the context

markfields replied:
Yeah I would love to pull serialization up higher.

markfields replied:
I'll take a look at that spot for tracking the viable contents too

markfields replied:
@anthony-murphy - I looked at moving to ChannelDeltaConnection, but decided not to in this PR.  I think I can actually get serialization all the way up to the ContainerRuntime, which would mean I could put the viable content straight into PendingStateManager and avoid any kind of tracking down below.

Also, when I looked into moving it to ChannelDeltaConnection I didn't immediately see how to get the source handle available to the ChannelDeltaConnection.

I'll have this problem with serializing in the ContainerRuntime too, I'm thinking about changing `makeHandlesSerializable` to replace the target handle with an object `{ from: this.handle, to: targetHandle }` so that whoever does the serialization can do the bind.  Haven't thought it through all the way yet, I wish there were a way to do it without mucking with the op at all. We can always talk about updating all the submit APIs I guess.

markfields committed:
Remove deferredHandle stuff

markfields committed:
Restore dirty state on exiting staging mode

markfields committed:
Try tracking staged v. submitted pending ops in DDS layer

This has some bugs:
- If detached when exiting staging mode, we don't flush the staged queue to the submitted queue
- When reconnecting, we don't preserve the staged v. not bit on the pending changes

Going to revert this because this stuff isn't necessary for the immediate scenario

markfields committed:
Revert "Try tracking staged v. submitted pending ops in DDS layer"

This reverts commit 11b8667f3e815a3afd6c09958d975c33bc8abe47.

markfields committed:
Pass viable contents (with handle to real object) to resubmit via wrapped localOpMetadata

markfields committed:
Missed unstaged change

markfields committed:
Fix two bugs

- wrappedLOM is undefined for non-local ops, so need to default to {} so it translates to undefined localOpMetadata properly
- I prematurely deleted a bunch of deprecated code

markfields committed:
Build fix

markfields committed:
Merge branch 'test/staging-mode' into sm/dist-psm

markfields committed:
nit

markfields committed:
undo unrelated change

markfields committed:
Tweak order of execution in exitStagingMode

markfields committed:
Revert unrelated changes

markfields commented on packages/dds/shared-object-base/src/sharedObject.ts:
```diff
@@ -458,10 +458,11 @@ export abstract class SharedObjectCore<
 	protected submitLocalMessage(content: unknown, localOpMetadata: unknown = undefined): void {
 		this.verifyNotClosed();
 		if (this.isAttached()) {
+			const wrappedLOM = { localOpMetadata, viableContent: content };
```
Because I can't move the `makeHandlesSerializabke` call there, because it didn't appear to have access to the source handle

markfields replied:
And I can't change the signature of the submit function because it's legacy-alpha

markfields replied:
I agree the asymmetry is not ideal, it's something I want to iterate on

markfields commented on packages/dds/shared-object-base/src/serializer.ts:
```diff
@@ -227,9 +216,7 @@ export class FluidSerializer implements IFluidSerializer {
-		if (this.runtime.inStagingMode === true) {
-			this.deferedHandleMap.set(handle.absolutePath, handle);
-		} else {
+		if (this.runtime.inStagingMode !== true) {
```
I'll keep looking at moving this check into bind later

markfields commented:
Obsolete.  Managed to keep the op un-encoded all the way up to ContainerRuntime without stashing it on LOM.  See #24230 and precursors

markfields commented:
For posterity: This PR started as an attempt at "Distributed Pending State", where each channel would hold its pending ops. But this got overly complex when trying to incorporate staging mode state, since each pending queue needs to be able to divide into pre-staging and staged changes.
